### PR TITLE
Handling promises at router layer

### PR DESCRIPTION
**Pull request by davidbayo10**
This feature will avoid try/catch block in all express controllers with async functions (async/await) or functions that returns a promise

### CODE REVIEW

**Comment by dougwilson** [comment]
Is this a duplicate of #47 ?

**Comment by davidbayo10** [comment]
Yes, I did not see it. Close if you want or tell me what PR is better in terms of code.

**Comment by dougwilson** [comment]
Ok. I'll review through your code to understand what it does. If you can provide highlights of the differences between the two PRs, that would be super helpful as well 👍 On the surface it seems the main difference is that #47 passes CI and this PR doesn't pass the CI.

**Comment by davidbayo10** [comment]
I think my code is simpliest and more readable. That's true, I have not passed the CI. But I don't know why, because I passed tests (`npm test`) before push. 😔

**Comment by dougwilson** [comment]
What is the behavior of this code for a promise that is rejected without a value (a valid rejection)? i.e. calling `reject()`

**Comment by davidbayo10** [comment]
Actually, an incorrect behavior. Fixing...

**Comment by davidbayo10** [comment]
image

This error comes from 2.0 branch. That's why builds are failing.

File: test/router.js [hunk 1]
```diff
@@ -124,6 +124,92 @@ describe('Router', function () {
...
+        promise
+          .then(function () {
+            saw(req, res)
+          })
+
+        return promise
+      })
+
+      router.use(function (err, req, res, next) {
+        res.sendStatus(500)
```
Comment (dougwilson) [inline test/router.js]: There is no such thing as res.sendStatus

File: test/router.js [hunk 2]
```diff
@@ -124,6 +124,92 @@ describe('Router', function () {
...
+          setTimeout(function () {
+            reject();
+          }, 100);
+        });
+
+        return promise
+      })
+
+      router.use(function (err, req, res, next) {
+        res.sendStatus(500)
```
Comment (dougwilson) [inline test/router.js]: There is no such thing as res.sendStatus

File: test/router.js [hunk 3]
```diff
@@ -124,6 +124,92 @@ describe('Router', function () {
       .expect(404, cb)
     })
 
+    it('should support promise as route handler', function (done) {
```
Comment (dougwilson) [inline test/router.js]: What does this test? I ran this test without any of your changes and it passed. This test doesn't seem to actually depend on any change, or maybe I'm just missing something.

File: test/router.js [hunk 4]
```diff
@@ -124,6 +124,92 @@ describe('Router', function () {
...
+          setTimeout(function () {
+            reject(new Error('async request failed'));
+          }, 100);
+        });
+
+        return promise
+      })
+
+      router.use(function (err, req, res, next) {
+        res.sendStatus(500)
```
Comment (dougwilson) [inline test/router.js]: There is no such thing as res.sendStatus

**Comment by davidbayo10** [comment]
Ok, I think both PR have same functionality then Imy PR doesn't add value. As you said, the other author have spent his time in development. I expect that you release 2.0 version soon. Thank you for all. :)

**Pull request closed without merging**